src/dmd/expressionsem.d

```diff
@@ -5521,13 +5524,87 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+            /*
+            {
+              auto a = e1, b = e2;
+              assert(a == b, _d_assert_fail!"=="(a, b));
+            }()
+            */
+            const tok = exp.e1.op;
+            bool isEqualsCallExpression;
+            if (global.params.checkAction == CHECKACTION.context && tok == TOK.call)
+            {
+                const callExp = cast(CallExp) exp.e1;
+                const callExpIdent = callExp.f.ident;
+                isEqualsCallExpression = callExpIdent == Id.__equals ||
+                                         callExpIdent == Id.eq;
+            }
+            if (global.params.checkAction == CHECKACTION.context &&
+                (tok == TOK.equal || tok == TOK.notEqual ||
+                 tok == TOK.lessThan || tok == TOK.greaterThan ||
+                 tok == TOK.lessOrEqual || tok == TOK.greaterOrEqual ||
+                 tok == TOK.identity || tok == TOK.notIdentity ||
+                 isEqualsCallExpression))
+            {
+                auto es = new Expressions();
+                auto tiargs = new Objects();
+                Loc loc = exp.e1.loc;
+
+                if (isEqualsCallExpression)
+                {
+                    auto callExp = cast(CallExp) exp.e1;
+                    auto args = callExp.arguments;
+
+                    // template args
+                    static immutable compMsg = "==";
+                    Expression comp = new StringExp(loc, cast(char*) compMsg.ptr);
+                    comp = comp.expressionSemantic(sc);
+                    tiargs.push(comp);
+                    tiargs.push((*args)[0].type);
+                    tiargs.push((*args)[1].type);
+
+                    // runtime args
+                    es.push((*args)[0]);
+                    es.push((*args)[1]);
+                }
+                else
+                {
+                    auto binExp = cast(EqualExp) exp.e1;
+
+                    // template args
+                    Expression comp = new StringExp(loc, cast(char*) Token.toChars(exp.e1.op));
+                    comp = comp.expressionSemantic(sc);
+                    tiargs.push(comp);
+                    tiargs.push(binExp.e1.type);
+                    tiargs.push(binExp.e2.type);
+
+                    // runtime args
+                    es.push(binExp.e1);
+                    es.push(binExp.e2);
+                }
+
+                Expression __assertFail = new IdentifierExp(exp.loc, Id.empty);
+                auto assertFail = new DotIdExp(loc, __assertFail, Id.object);
+
+                auto dt = new DotTemplateInstanceExp(loc, assertFail, Id._d_assert_fail, tiargs);
+                auto ec = CallExp.create(Loc.initial, dt, es);
+                exp.msg = ec;
+            }
+            else
+            {
+                OutBuffer buf;
+                buf.printf("%s failed", assertExpMsg);
+                exp.msg = new StringExp(Loc.initial, buf.extractString());
```

Comment:
This originates from https://github.com/dlang/dmd/pull/7575, but leads to the iasm failures on 32-bit as foo isn't on EBP anymore:

```diff
 Disassembly of section .text._D3foo5test1FZv:
@@ -7,23 +7,31 @@ Disassembly of section .text._D3foo5test
 00000000 <_D3foo5test1FZv>:
    0:	55                   	push   ebp
    1:	8b ec                	mov    ebp,esp
-   3:	83 ec 08             	sub    esp,0x8
-   6:	31 c0                	xor    eax,eax
-   8:	89 45 f8             	mov    DWORD PTR [ebp-0x8],eax
-   b:	89 45 fc             	mov    DWORD PTR [ebp-0x4],eax
-   e:	90                   	nop
-   f:	90                   	nop
-  10:	b8 08 00 00 00       	mov    eax,0x8
-  15:	89 45 f8             	mov    DWORD PTR [ebp-0x8],eax
-  18:	83 7d f8 08          	cmp    DWORD PTR [ebp-0x8],0x8
-  1c:	74 0d                	je     2b <_D3foo5test1FZv+0x2b>
-  1e:	6a 0f                	push   0xf
-  20:	b9 00 00 00 00       	mov    ecx,0x0
-  25:	51                   	push   ecx
-  26:	e8 fc ff ff ff       	call   27 <_D3foo5test1FZv+0x27>
-  2b:	c9                   	leave  
-  2c:	c3                   	ret    
-  2d:	00 00                	add    BYTE PTR [eax],al
+   3:	83 ec 10             	sub    esp,0x10
+   6:	89 5d f0             	mov    DWORD PTR [ebp-0x10],ebx
+   9:	89 75 f4             	mov    DWORD PTR [ebp-0xc],esi
+   c:	31 c0                	xor    eax,eax
+   e:	89 45 f8             	mov    DWORD PTR [ebp-0x8],eax
+  11:	89 45 fc             	mov    DWORD PTR [ebp-0x4],eax
+  14:	b8 10 00 00 00       	mov    eax,0x10
+  19:	89 45 f8             	mov    DWORD PTR [ebp-0x8],eax
+  1c:	83 7d f8 08          	cmp    DWORD PTR [ebp-0x8],0x8
+  20:	74 1f                	je     41 <_D3foo5test1FZv+0x41>
+  22:	6a 0f                	push   0xf
+  24:	b9 00 00 00 00       	mov    ecx,0x0
+  29:	bb 05 00 00 00       	mov    ebx,0x5
+  2e:	51                   	push   ecx
+  2f:	53                   	push   ebx
+  30:	ba 06 00 00 00       	mov    edx,0x6
+  35:	be 17 00 00 00       	mov    esi,0x17
+  3a:	52                   	push   edx
+  3b:	56                   	push   esi
+  3c:	e8 fc ff ff ff       	call   3d <_D3foo5test1FZv+0x3d>
+  41:	8b 5d f0             	mov    ebx,DWORD PTR [ebp-0x10]
+  44:	8b 75 f4             	mov    esi,DWORD PTR [ebp-0xc]
+  47:	c9                   	leave  
+  48:	c3                   	ret    
+  49:	00 00                	add    BYTE PTR [eax],al
```

Are we supposed to cater for this?